mlir/lib/Quantum/Transforms/loop_boundary_optimization.cpp

```diff
@@ -0,0 +1,76 @@
+// Copyright 2025 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "loop-boundary"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/Debug.h"
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/RegionKindInterface.h"
+#include "mlir/Interfaces/ControlFlowInterfaces.h"
+#include "mlir/Interfaces/LoopLikeInterface.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace catalyst {
+namespace quantum {
+
+#define GEN_PASS_DEF_LOOPBOUNDARYOPTIMIZATIONPASS
+#define GEN_PASS_DECL_LOOPBOUNDARYOPTIMIZATIONPASS
+#include "Quantum/Transforms/Passes.h.inc"
+
+struct LoopBoundaryOptimizationPass
```

Comment:
This doesn't have to be solved now, but I do think in the future we'll want a single pass that applies many quantum peephole optimizations, similarly to the canonicalize pass in MLIR. That way, we can actually make use of the fixed-point greedy pattern applicator, taking advantage of one peephole optimization enabling another.